### ISSUE TITLE
[2020 NeurIPS] Supervised Contrastive Learning

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#40]
Supervised cross-entropy loss 는 noisy labels에 robust하지 않고 possibility of poor margins로 인해 generalization performance가 저하된다는 단점이 있음

Self-supervised learning에서 사용되고 있는 contrastive learning을 supervised task에서 사용할 수 있도록 `supervised contrastive (SupCon) loss`를 설계한 논문

## 1. Main Contribution
- SupCon loss performs better than naive extension of self-supervised contrastive loss
- SupCon gradient encourages learning from hard positives and hard negatives
- SupCon is more better than cross-entropy to top-1 accuracy for a number of datasets
- SupCon is more robust than cross-entropy to natural corruptions
- SupCon is less sensitive than cross-entropy to a range of hyperparameters

## 2. Contrastive Learning
![image](https://user-images.githubusercontent.com/49643709/138789802-966a87af-26bc-42f0-9782-71f152d35242.png)
![image](https://user-images.githubusercontent.com/49643709/138789737-4cbd63a7-0507-45ad-88c0-b21f6f16ad79.png)

#### 2.1.1. 2 types of loss
- Triplet loss : one postivie and one negative
- N-pairs loss : one positive and many negatives

#### 2.1.2. Self-supervised contrastive loss
![image](https://user-images.githubusercontent.com/49643709/139245594-609a0a1f-aa4d-4ff4-bf22-9c2f7aa83fae.png)
- For each anchor i, there is 1 positive pair and 2N - 2 negative pairs
- Denominator consists of 2N - 1 terms (positive and negatives)

####  2.2.1. 3 main components
1. Data augmentation module `Aug( )`
    - For each input sample `x`, generate 2 random augmentations 'Aug(x)`
    - Each `Aug(x)` represents a different view of the data and contains some subset of the information in the original sample
2. Encoder network module `Enc( )`
    - Maps `Aug(x)` to a representation `r = Enc(Aug(x))`
    - `r` is `normalized` to the unit hypersphere (encoder dimension = 2048)
3. Projection network `Proj( )`
    - Maps `r` to a vector `z = Proj(r)`
    - `z` if `normalized` to the unit hypersphere (projection dimension = 128)
    - Use inner product to measure distances
    - Discard `Proj( )` at the end of contrastive learning

#### 2.2.2. Supervised contrastive loss
![image](https://user-images.githubusercontent.com/49643709/139245728-b0fe79a2-74e8-4884-a26a-f5b2f75db8a9.png)
- `P(i)` is the set of indices of all positives in the multiviewed batch distinct from i
- `|P(i)|` is cardinality

#### 2.2.3. 3 desirable loss properties
1. Generalization to an arbitrary number of positives
    - All positives contribute to the numerator
    - More robust clustering of the representation space than self-supervised contrastive loss cause of closely aligned representations to all entries from the same class
2. Contrastive power increases with more negatives
    - Ability to discriminate between signal and noise (negatives) is improved by adding more examples of negatives
3. Intrinsic ability to perform hard positive/negative mining
    - Normalized representations induces a gradient structure that gives rise to implicit hard positive/negative mining
    - The gradient contributions from hard positives/negatives are large while those for easy positives/negatives are small

1번과 3번의 details는 해당 논문의 supplementary 참고

#### 2.2.4. Which loss to use
<img src='https://user-images.githubusercontent.com/49643709/139472513-ad590c70-879b-4742-a48a-6a0ff6598e4c.png' width=60% height=60%>

Two loss formutations are not equivalent, which `L_out` performs better

For `L_out`, the positives normalization factor remove bias present in the positives in a multiviewed batch

Though `L_in` contains the normalization factor inside of the log, which does not affect the gradient

Use `L_out` for experiments

이에 대한 details는 해당 논문의 supplementary 참고

### 3.1. Classification Accuracy
![image](https://user-images.githubusercontent.com/49643709/139593281-109b93b7-2144-45b3-9482-88b9c227686f.png)
![image](https://user-images.githubusercontent.com/49643709/139593287-813cffe1-b40e-4191-9e2a-616b32b5885a.png)
- SupCon가 self-supervised contrastive loss보다 더 우수함
- SupCon가 supervised cross-entropy loss보다 더 우수함

### 3.2. Robustness to Image Corruptions and Reduced Training Data
![image](https://user-images.githubusercontent.com/49643709/139593336-36669618-ba5e-4864-ab47-1cdd1db3e8a5.png)
- SupCon가 cross-entropy보다 더 우수함

![image](https://user-images.githubusercontent.com/49643709/139799819-acbfe0ad-7e9f-4948-b8c0-fdb57d4c2ec6.png)
![image](https://user-images.githubusercontent.com/49643709/139799881-447601bd-9ff3-46d0-b8ca-9e825134c65a.png)
- ECE score를 보면 corruption severity가 높아질 수록 cross entropy가 우수한 경우가 있음 (ResNet-200의 경우)
- Corruption severity가 높아지면 natural scene에서 보기 힘들고 cross entropy보다 성능이 낮기에 `natural corruption에 robust하다`는 표현이 서 해당 워딩을 사용하지 않은게 아닐까 생각됨

### 3.3. Hyperparameter Stability
![image](https://user-images.githubusercontent.com/49643709/139593395-7413af06-99c9-4018-a733-bc856c89ee33.png)
- SupCon가 cross-entropy보다 hyperparameter stability가 우수함 (augmentation, optimizer, learning rate)
- Low temperature를 사용하는 것이 high temperature보다 우수하나, extremely low temperature를 사용하면 numerical instability 때문에 학습이 어려워짐

### 3.4. Transfer Leaning
![image](https://user-images.githubusercontent.com/49643709/139593523-159a660d-aae0-4641-a9dc-ff2e31a950bd.png)
- SupCon이 다른 모델들보다 우수하지 않음
